ports/esp32/main.c

```diff
@@ -96,6 +96,25 @@ time_t platform_mbedtls_time(time_t *timer) {
     return tv.tv_sec + TIMEUTILS_SECONDS_1970_TO_2000;
 }
 
+// Custom mbedTLS allocator function
+static void *tls_calloc(size_t c, size_t len) {
+    // Calling plain calloc() allows mbedTLS to allocate from PSRAM or internal memory
+    //
+    // (The ESP-IDF default is internal-only, partly for physical security to prevent
+    // possible information leakage from unencrypted PSRAM contents on the original
+    // ESP32 - no PSRAM encryption on that chip. MicroPython doesn't support flash
+    // encryption and is already storing the Python heap in PSRAM so this isn't a
+    // significant factor in overall physical security.)
+    void *p = calloc(c, len);
+    if (p == NULL) {
+        // A Python object (i.e. SSL socket) may be holding references
+        // to some ESP-IDF buffers which will be freed after collection
+        gc_collect();
```

Comment:
Are we really 100% certain that this `tls_calloc` function is only ever called from within MicroPython context (ie GIL is held), and so it's safe to call `gc_collect` here?

(I'm thinking about things like mDNS services that run in the background independently from the MP thread and may need to make mbedtls calls, which may then end up here.  Also maybe user C modules that start a C thread that uses TLS.)

Reply:
mDNS is all plaintext, isn't it?

But I think the user C module case is a very strong one, will refactor to check the current thread is a Python thread.

Reply:
Have rewritten to handle only the special case of `mbedtls_ssl_setup()` as this is the function most likely to fail to allocate.

I think there is a possible generic version of this (maybe applicable to all C-level mallocs), but need to:

* Make GIL mutex recursive on esp32 so we can take it while calling gc_collect().
* Handle non-Python threads (either by skipping GC on these, or by doing what the BLE stack does and making a temporary fake Python thread context.)

But this way is much smaller scoped change, covers the most obvious failure pattern.

Reply:
> Have rewritten to handle only the special case of `mbedtls_ssl_setup()` as this is the function most likely to fail to allocate.

That is a lot cleaner and more generic (not tied to esp32).



> * Make GIL mutex recursive on esp32 so we can take it while calling gc_collect().

I don't think a recursive GIL is necessary.  See `extmod/modbluetooth.c` and grep for `MICROPY_PY_BLUETOOTH_USE_SYNC_EVENTS_WITH_INTERLOCK`: it only obtains the GIL if it's not already within Python context.

Reply:
> I don't think a recursive GIL is necessary.

I was thinking of protecting against the case where a user C module releases the GIL on a Python thread. But maybe this isn't something a user C module is able to do?

Reply:
> I was thinking of protecting against the case where a user C module releases the GIL on a Python thread

Yes, a user C module can definitely do that.

So I think what you have now is a good solution.